src/main/java/org/broadinstitute/hellbender/engine/ReadInput.java

```diff
@@ -0,0 +1,170 @@
+import org.broadinstitute.hellbender.utils.Utils;
+
+import java.io.File;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Class to represent a Read-containing input file. Tools should declare @Argument-annotated fields of
+ * this type (or Collections of this type), and the engine will automatically discover
+ * them at runtime.
+ *
+ * DO NOT ATTEMPT TO INSTANTIATE THIS CLASS DIRECTLY! ReadInputs must be instantiated by the argument-parsing
+ * system only. This is why the constructor is marked as protected.
+ *
+ * ReadInputs can be assigned logical names on the command line using the syntax:
+ *
+ *     --argument_name logical_name:read_file
+ *
+ * These logical names can then be retrieved by the tool at runtime via {@link #getName}
+ *
+ * Furthermore, a list of comma-separated key=value pairs may be provided as follows:
+ *
+ *     --argument_name logical_name,key1=value1,key2=value2:read_file
+ *
+ * the string value provided for a given key can be retrieved via {@link #getAttribute(String)}. Keys must be unique.
+ *
+ */
+public final class ReadInput {
+
+    /**
+     * Logical name for this source of Features optionally provided by the user on the command line
+     * using the --argument_name logical_name:read_file syntax. Defaults to the absolute path of
+     * the underlying file if no logical name is specified
+     */
+    private final String name;
+
+    private final Map<String, String> kevValueMap;
+
+    /**
+     * File containing Reads as specified by the user on the command line
+     */
+    private final File readFile;
+
+    /**
+     * Construct a ReadInput from a String argument value either of the form "logical_name:read_file"
+     * or simply "read_file".
+     *
+     * Only meant to be called by the argument parsing system, and therefore marked as package-visible --
+     * ReadInputs constructed some other way will not be recognized by the engine.
+     *
+     * Note: cannot delegate to another constructor because Java only allows a call to "this" on the first line of a constructor.
+     *
+     * @param rawArgumentValue String of the form "logical_name:read_file" or "read_file"
+     */
+    ReadInput(final String rawArgumentValue) {
+        Utils.nonNull(rawArgumentValue, "rawArgumentValue");
+        final ParsedArgument parsedArgument = ParsedArgument.of(rawArgumentValue);
+
+        name = parsedArgument.getName();
+        kevValueMap = parsedArgument.keyValueMap();
+        readFile = parsedArgument.getFile();
+    }
+
+    /**
+     * Construct a ReadInput from raw components: name, key value pairs and the file.
+     *
+     * This constructor is meant to be called by the engine and test classes --
+     * ReadInputs constructed some other way will not be recognized by the engine.
+     */
+    @VisibleForTesting
+    public ReadInput(final String name, final Map<String, String> kevValueMap, final File readFile) {
```

Comment:
If it's visible for testing, make it package-visible, and put the test in the same package